File: source/clibCore/Interfaces/cpApi/CPUserClass.vb

Option Explicit On
Option Strict On

Imports Contensive.Core.Controllers
Imports Contensive.Core.Controllers.genericController
Imports Contensive.BaseClasses
Imports System.Runtime.InteropServices

Namespace Contensive.Core
    '
    ' comVisible to be activeScript compatible
    '
    <ComVisible(True)>
    <ComClass(CPUserClass.ClassId, CPUserClass.InterfaceId, CPUserClass.EventsId)>
    Public Class CPUserClass
        Inherits BaseClasses.CPUserBaseClass
        Implements IDisposable
        '
#Region "COM GUIDs"
        Public Const ClassId As String = "08EF64C6-9C51-4B32-84D9-0D3BDAF42A28"
        Public Const InterfaceId As String = "B1A95B1F-A00D-4AC6-B3A0-B1619568C2EA"
        Public Const EventsId As String = "DBE2B6CB-6339-4FFB-92D7-BE37AEA841CC"
#End Region
        '
        Private cpCore As Contensive.Core.coreClass
        Private CP As CPClass
        Protected disposed As Boolean = False
        '
        '====================================================================================================
        '
        Public Sub New(ByVal cpCoreObj As Contensive.Core.coreClass, ByVal CPParent As CPClass)
            MyBase.New()
            cpCore = cpCoreObj
            CP = CPParent
        End Sub
        '
        '====================================================================================================
        '
        Protected Overridable Overloads Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposed Then
                Call appendDebugLog(".dispose, dereference cp, main, csv")
                If disposing Then
                    '
                    ' call .dispose for managed objects
                    '
                    CP = Nothing
                    cpCore = Nothing
                End If
                '
                ' Add code here to release the unmanaged resource.
                '
            End If
            Me.disposed = True
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property Email() As String 'Inherits BaseClasses.CPUserBaseClass.Email
            Get
                If True Then
                    Return CP.core.authContext.user.Email
                Else
                    Return ""
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function GetIdByLogin(ByVal Username As String, ByVal Password As String) As Integer 'Inherits BaseClasses.CPUserBaseClass.GetIdByLogin
            If True Then
                Return CP.core.authContext.authenticateGetId(cpCore, Username, Password)
            Else
                Return 0
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property Id() As Integer 'Inherits BaseClasses.CPUserBaseClass.Id
            Get
                Dim localId As Integer = 0
                '
                If True Then
                    localId = CP.core.authContext.user.ID
                    If (localId = 0) Then
                        localId = CP.core.db.metaData_InsertContentRecordGetID("people", 0)
                        Call CP.core.authContext.recognizeById(cpCore, localId, CP.core.authContext)
                    End If
                End If
                Return localId
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsAdmin() As Boolean 'Inherits BaseClasses.CPUserBaseClass.admin
            Get
                If True Then
                    Return CP.core.authContext.isAuthenticatedAdmin(cpCore)
                Else
                    Return False
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function IsAdvancedEditing(ByVal ContentName As String) As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsAdvancedEditing
            If True Then
                Return CP.core.authContext.isAdvancedEditing(cpCore, ContentName)
            Else
                Return False
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsAuthenticated() As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsAuthenticated
            Get
                If True Then
                    Return (CP.core.authContext.isAuthenticated())
                Else
                    Return False
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function IsAuthoring(ByVal ContentName As String) As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsAuthoring
            If True Then
                Return CP.core.authContext.isEditing(cpCore, ContentName)
            Else
                Return False
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsContentManager(Optional ByVal ContentName As String = "Page Content") As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsContentManager
            If True Then
                Return CP.core.authContext.isAuthenticatedContentManager(cpCore, ContentName)
            Else
                Return False
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsDeveloper() As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsDeveloper
            Get
                If True Then
                    Return CP.core.authContext.isAuthenticatedDeveloper(cpCore)
                Else
                    Return False
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function IsEditing(ByVal ContentName As String) As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsEditing
            If True Then
                Return CP.core.authContext.isEditing(cpCore, ContentName)
            Else
                Return False
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsEditingAnything() As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsEditingAnything
            Get
                If True Then
                    Return CP.core.authContext.isEditingAnything(cpCore)
                Else
                    Return False
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsGuest() As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsGuest
            Get
                If True Then
                    Return CP.core.authContext.isGuest(cpCore)
                Else
                    Return False
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function IsInGroup(ByVal groupName As String, Optional ByVal userId As Integer = 0) As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsInGroup
            Dim groupId As Integer
            Dim result As Boolean
            '
            Try
                groupId = CP.Group.GetId(groupName)
                If userId = 0 Then
                    userId = Id
                End If
                If groupId = 0 Then
                    result = False
                Else
                    result = IsInGroupList(groupId.ToString, userId)
                End If
            Catch ex As Exception
                Call CP.core.handleExceptionAndContinue(ex, "Unexpected error in cs.user.IsInGroup")
                result = False
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsInGroupList(ByVal GroupIDList As String, Optional ByVal userId As Integer = 0) As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsInGroup
            Dim result As Boolean
            '
            Try
                If userId = 0 Then
                    userId = Id
                End If
                result = CP.core.authContext.isMemberOfGroupIdList(cpCore, userId, IsAuthenticated(), GroupIDList, False)
            Catch ex As Exception
                Call CP.core.handleExceptionAndContinue(ex, "Unexpected error in cs.user.IsInGroupList")
                result = False
            End Try
            Return result
            '
        End Function
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsMember() As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsMember
            Get
                If True Then
                    Return CP.core.authContext.isAuthenticatedMember(cpCore)
                Else
                    Return False
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function IsQuickEditing(ByVal ContentName As String) As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsQuickEditing
            If True Then
                Return CP.core.authContext.isQuickEditing(cpCore, ContentName)
            Else
                Return False
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsRecognized() As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsRecognized
            Get
                If True Then
                    Return CP.core.authContext.isRecognized(cpCore)
                Else
                    Return False
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsWorkflowRendering() As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsWorkflowRendering
            Get
                If True Then
                    Return CP.core.pages.pagemanager_IsWorkflowRendering
                Else
                    Return False
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property Language() As String 'Inherits BaseClasses.CPUserBaseClass.Language
            Get
                If True Then
                    Return CP.core.authContext.user.language
                Else
                    Return ""
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property LanguageID() As Integer 'Inherits BaseClasses.CPUserBaseClass.LanguageId
            Get
                If True Then
                    Return CP.core.authContext.user.LanguageID
                Else
                    Return 0
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function Login(ByVal UsernameOrEmail As String, ByVal Password As String, Optional ByVal SetAutoLogin As Boolean = False) As Boolean 'Inherits BaseClasses.CPUserBaseClass.Login
            If True Then
                Return CP.core.authContext.authenticate(cpCore, UsernameOrEmail, Password, SetAutoLogin)
            Else
                Return False
            End If
        End Function
        '
        '====================================================================================================
        '
        <Obsolete("Use LoginById(integer) instead", False)>
        Public Overrides Function LoginByID(ByVal RecordID As String, Optional ByVal SetAutoLogin As Boolean = False) As Boolean
            Return CP.core.authContext.authenticateById(cpCore, EncodeInteger(RecordID), CP.core.authContext)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function LoginByID(ByVal RecordID As Integer) As Boolean
            Return CP.core.authContext.authenticateById(cpCore, RecordID, CP.core.authContext)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function LoginByID(ByVal RecordID As Integer, ByVal SetAutoLogin As Boolean) As Boolean
            Return CP.core.authContext.authenticateById(cpCore, RecordID, CP.core.authContext)
            If Not CP.core.authContext.user.AutoLogin Then
                CP.core.authContext.user.AutoLogin = True
                CP.core.authContext.user.saveObject(cpCore)
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function LoginIsOK(ByVal UsernameOrEmail As String, ByVal Password As String) As Boolean 'Inherits BaseClasses.CPUserBaseClass.LoginIsOK
            If True Then
                Return CP.core.authContext.main_IsLoginOK(cpCore, UsernameOrEmail, Password)
            Else
                Return False
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Sub Logout() 'Inherits BaseClasses.CPUserBaseClass.Logout
            Call CP.core.authContext.logout(cpCore)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property Name() As String 'Inherits BaseClasses.CPUserBaseClass.Name
            Get
                If True Then
                    Return CP.core.authContext.user.Name
                Else
                    Return ""
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property IsNew() As Boolean 'Inherits BaseClasses.CPUserBaseClass.IsNew
            Get
                Return CP.core.authContext.visit.MemberNew
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function IsNewLoginOK(ByVal Username As String, ByVal Password As String) As Boolean 'Inherits BaseClasses.CPUserBaseClass.NewLoginIsOK
            Dim errorMessage As String = ""
            Dim errorCode As Integer = 0
            Return CP.core.authContext.isNewLoginOK(cpCore, Username, Password, errorMessage, errorCode)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property OrganizationID() As Integer 'Inherits BaseClasses.CPUserBaseClass.OrganizationId
            Get
                If True Then
                    Return CP.core.authContext.user.OrganizationID
                Else
                    Return 0
                End If
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property Password() As String 'Inherits BaseClasses.CPUserBaseClass.Password
            Get
                Return CP.core.authContext.user.Password
            End Get
        End Property
        '
        '====================================================================================================
        '
        Public Overrides Function Recognize(ByVal UserID As Integer) As Boolean 'Inherits BaseClasses.CPUserBaseClass.Recognize
            Return CP.core.authContext.recognizeById(cpCore, UserID, CP.core.authContext)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides ReadOnly Property Username() As String 'Inherits BaseClasses.CPUserBaseClass.Username
            Get
                Return CP.core.authContext.user.Username
            End Get
        End Property
        '
        '=======================================================================================================
        '
        Public Overrides Function GetProperty(ByVal PropertyName As String, Optional ByVal DefaultValue As String = "", Optional ByVal TargetMemberId As Integer = 0) As String
            If (TargetMemberId = 0) Then
                Return CP.core.userProperty.getText(PropertyName, DefaultValue)
            Else
                Return CP.core.userProperty.getText(PropertyName, DefaultValue, TargetMemberId)
            End If
        End Function
        '
        '=======================================================================================================
        '
        Public Overrides Sub SetProperty(ByVal PropertyName As String, ByVal Value As String, Optional ByVal TargetMemberId As Integer = 0)
            If (TargetMemberId = 0) Then
                Call CP.core.userProperty.setProperty(PropertyName, Value)
            Else
                Call CP.core.userProperty.setProperty(PropertyName, Value, TargetMemberId)
            End If
        End Sub
        '
        '=======================================================================================================
        ' REFACTOR -- obsolete this and setup the defaultValue type correctly, and add the targetUserId
        '=======================================================================================================
        '
        Public Overrides Function GetBoolean(PropertyName As String, Optional DefaultValue As String = "") As Boolean
            Return CP.core.userProperty.getBoolean(PropertyName, genericController.EncodeBoolean(DefaultValue))
        End Function
        '
        '=======================================================================================================
        ' REFACTOR -- obsolete this and setup the defaultValue type correctly, and add the targetUserId
        '=======================================================================================================
        '
        Public Overrides Function GetDate(PropertyName As String, Optional DefaultValue As String = "") As Date
            Return CP.core.userProperty.getDate(PropertyName, genericController.EncodeDate(DefaultValue))
        End Function
        '
        '=======================================================================================================
        ' REFACTOR -- obsolete this and setup the defaultValue type correctly, and add the targetUserId
        '=======================================================================================================
        '
        Public Overrides Function GetInteger(PropertyName As String, Optional DefaultValue As String = "") As Integer
            Return CP.core.userProperty.getInteger(PropertyName, genericController.EncodeInteger(DefaultValue))
        End Function
        '
        '=======================================================================================================
        ' REFACTOR -- obsolete this and setup the defaultValue type correctly, and add the targetUserId
        '=======================================================================================================
        '
        Public Overrides Function GetNumber(PropertyName As String, Optional DefaultValue As String = "") As Double
            Return CP.core.userProperty.getNumber(PropertyName, EncodeNumber(DefaultValue))
        End Function
        '
        '=======================================================================================================
        '
        Public Overrides Function GetText(PropertyName As String, Optional DefaultValue As String = "") As String
            Return CP.core.userProperty.getText(PropertyName, DefaultValue)
        End Function
        '
        '====================================================================================================
        ' REFACTOR -- figure out what this does and rewrite
        '
        Public Overrides Sub track()
            Dim localId As Integer
            If True Then
                '
                ' get the id property, which triggers a track() if it returns o
                '
                localId = Id
                'If Id = 0 Then
                '    localId = cmc.csv_InsertContentRecordGetID("people", 0)
                '    Call cmc.main_RecognizeMemberByID(localId)
                'End If
            End If
        End Sub
        '
        '====================================================================================================
        '
        Private Sub appendDebugLog(ByVal copy As String)
            '
        End Sub
        '
        '====================================================================================================
        '
        Private Sub tp(ByVal msg As String)
            'Call appendDebugLog(msg)
        End Sub
#Region " IDisposable Support "
        ' Do not change or add Overridable to these methods.
        ' Put cleanup code in Dispose(ByVal disposing As Boolean).
        Public Overloads Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
        Protected Overrides Sub Finalize()
            Dispose(False)
            MyBase.Finalize()
        End Sub
#End Region
    End Class

End Namespace

File: source/core/Views/cpApi/CPUserClass.cs
﻿
using System;
using System.Reflection;
using System.Xml;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using Contensive.Core;
using Contensive.Core.Models.Entity;
using Contensive.Core.Controllers;
using static Contensive.Core.Controllers.genericController;
using static Contensive.Core.constants;
using Contensive.Core.Models.Context;
//
namespace Contensive.Core {
    public class CPUserClass : BaseClasses.CPUserBaseClass, IDisposable {
        //
        #region COM GUIDs
        public const string ClassId = "08EF64C6-9C51-4B32-84D9-0D3BDAF42A28";
        public const string InterfaceId = "B1A95B1F-A00D-4AC6-B3A0-B1619568C2EA";
        public const string EventsId = "DBE2B6CB-6339-4FFB-92D7-BE37AEA841CC";
        #endregion
        //
        private Contensive.Core.coreClass cpCore;
        private CPClass CP;
        protected bool disposed = false;
        //
        //====================================================================================================
        //
        public CPUserClass(Contensive.Core.coreClass cpCoreObj, CPClass CPParent) : base() {
            cpCore = cpCoreObj;
            CP = CPParent;
        }
        //
        //====================================================================================================
        //
        protected virtual void Dispose(bool disposing) {
            if (!this.disposed) {
                appendDebugLog(".dispose, dereference cp, main, csv");
                if (disposing) {
                    //
                    // call .dispose for managed objects
                    //
                    CP = null;
                    cpCore = null;
                }
                //
                // Add code here to release the unmanaged resource.
                //
            }
            this.disposed = true;
        }
        //
        //====================================================================================================
        //
        public override string Email
        {
            get {
                return CP.core.doc.sessionContext.user.Email;
            }
        }
        //
        //====================================================================================================
        //
        public override int GetIdByLogin(string Username, string Password) {
            return CP.core.doc.sessionContext.authenticateGetId(cpCore, Username, Password);
        }
        //
        //====================================================================================================
        /// <summary>
        /// Returns the id of the user in the current session context. If 0, this action will create a user.
        /// This trigger allows sessions with guest detection disabled that will enable if used.
        /// </summary>
        public override int Id {
            get {
                int localId = CP.core.doc.sessionContext.user.id;
                if (localId == 0) {
                    localId = CP.core.db.insertContentRecordGetID("people", 0);
                    CP.core.doc.sessionContext.recognizeById(cpCore, localId, ref CP.core.doc.sessionContext);
                }
                return localId;
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsAdmin {
            get {
                return CP.core.doc.sessionContext.isAuthenticatedAdmin(cpCore);
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsAdvancedEditing(string ContentName) {
            return CP.core.doc.sessionContext.isAdvancedEditing(cpCore, ContentName);
        }
        //
        //====================================================================================================
        //
        public override bool IsAuthenticated {
            get {
                return (CP.core.doc.sessionContext.isAuthenticated);
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsAuthoring(string ContentName) {
            return CP.core.doc.sessionContext.isEditing(ContentName);
        }
        //
        //====================================================================================================
        //
        public override bool IsContentManager(string ContentName = "Page Content") {
            return CP.core.doc.sessionContext.isAuthenticatedContentManager(cpCore, ContentName);
        }
        //
        //====================================================================================================
        //
        public override bool IsDeveloper {
            get {
                return CP.core.doc.sessionContext.isAuthenticatedDeveloper(cpCore);
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsEditing(string ContentName) {
            return CP.core.doc.sessionContext.isEditing(ContentName);
        }
        //
        //====================================================================================================
        //
        public override bool IsEditingAnything {
            get {
                return CP.core.doc.sessionContext.isEditingAnything();
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsGuest {
            get {
                return CP.core.doc.sessionContext.isGuest(cpCore);
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsInGroup(string groupName, int userId = 0) {
            int groupId = 0;
            bool result = false;
            //
            try {
                groupId = CP.Group.GetId(groupName);
                if (userId == 0) {
                    userId = Id;
                }
                if (groupId == 0) {
                    result = false;
                } else {
                    result = IsInGroupList(groupId.ToString(), userId);
                }
            } catch (Exception ex) {
                CP.core.handleException(ex);
                result = false;
            }
            return result;
        }
        //
        //====================================================================================================
        //
        public override bool IsInGroupList(string GroupIDList, int userId = 0) //Inherits BaseClasses.CPUserBaseClass.IsInGroup
        {
            bool result = false;
            //
            try {
                if (userId == 0) {
                    userId = Id;
                }
                result = CP.core.doc.sessionContext.isMemberOfGroupIdList(cpCore, userId, IsAuthenticated, GroupIDList, false);
            } catch (Exception ex) {
                CP.core.handleException(ex);
                result = false;
            }
            return result;
            //
        }
        //
        //====================================================================================================
        //
        public override bool IsMember {
            get {
                return CP.core.doc.sessionContext.isAuthenticatedMember(cpCore);
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsQuickEditing(string ContentName) {
            return CP.core.doc.sessionContext.isQuickEditing(cpCore, ContentName);
        }
        //
        //====================================================================================================
        //
        public override bool IsRecognized {
            get {
                return CP.core.doc.sessionContext.isRecognized(cpCore);
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsWorkflowRendering {
            get {
                return CP.core.doc.sessionContext.isWorkflowRendering();
            }
        }
        //
        //====================================================================================================
        //
        public override string Language {
            get {
                if (CP.core.doc.sessionContext.userLanguage != null) {
                    return CP.core.doc.sessionContext.userLanguage.name;
                }
                return string.Empty;
            }
        }
        //
        //====================================================================================================
        //
        public override int LanguageID {
            get {
                return CP.core.doc.sessionContext.user.LanguageID;
            }
        }
        //
        //====================================================================================================
        //
        public override bool Login(string UsernameOrEmail, string Password, bool SetAutoLogin = false) {
            return CP.core.doc.sessionContext.authenticate(cpCore, UsernameOrEmail, Password, SetAutoLogin);
        }
        //
        //====================================================================================================
        //
        [Obsolete("Use LoginById(integer) instead", false)]
        public override bool LoginByID(string RecordID, bool SetAutoLogin = false) {
            return CP.core.doc.sessionContext.authenticateById(cpCore, encodeInteger(RecordID), CP.core.doc.sessionContext);
        }
        //
        //====================================================================================================
        //
        public override bool LoginByID(int RecordID) {
            return CP.core.doc.sessionContext.authenticateById(cpCore, RecordID, CP.core.doc.sessionContext);
        }
        //
        //====================================================================================================
        //
        public override bool LoginByID(int RecordID, bool SetAutoLogin) {
            bool result = CP.core.doc.sessionContext.authenticateById(cpCore, RecordID, CP.core.doc.sessionContext);
            if (result) {
                CP.core.doc.sessionContext.user.AutoLogin = SetAutoLogin;
                CP.core.doc.sessionContext.user.save(cpCore);
            }
            return result;
        }
        //
        //====================================================================================================
        //
        public override bool LoginIsOK(string UsernameOrEmail, string Password) {
            return CP.core.doc.sessionContext.isLoginOK(cpCore, UsernameOrEmail, Password);
        }
        //
        //====================================================================================================
        //
        public override void Logout() {
            CP.core.doc.sessionContext.logout(cpCore);
        }
        //
        //====================================================================================================
        //
        public override string Name {
            get {
                return CP.core.doc.sessionContext.user.name;
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsNew {
            get {
                return CP.core.doc.sessionContext.visit.MemberNew;
            }
        }
        //
        //====================================================================================================
        //
        public override bool IsNewLoginOK(string Username, string Password) {
            string errorMessage = "";
            int errorCode = 0;
            return CP.core.doc.sessionContext.isNewLoginOK(cpCore, Username, Password, ref errorMessage, ref errorCode);
        }
        //
        //====================================================================================================
        //
        public override int OrganizationID {
            get {
                return CP.core.doc.sessionContext.user.OrganizationID;
            }
        }
        //
        //====================================================================================================
        //
        public override string Password {
            get {
                return CP.core.doc.sessionContext.user.Password;
            }
        }
        //
        //====================================================================================================
        //
        public override bool Recognize(int UserID) {
            sessionContextModel authContext = CP.core.doc.sessionContext;
            return CP.core.doc.sessionContext.recognizeById(cpCore, UserID, ref authContext);
        }
        //
        //====================================================================================================
        //
        public override string Username {
            get {
                return CP.core.doc.sessionContext.user.Username;
            }
        }
        //
        //=======================================================================================================
        //
        public override string GetProperty(string PropertyName, string DefaultValue = "", int TargetMemberId = 0) {
            if (TargetMemberId == 0) {
                return CP.core.userProperty.getText(PropertyName, DefaultValue);
            } else {
                return CP.core.userProperty.getText(PropertyName, DefaultValue, TargetMemberId);
            }
        }
        //
        //=======================================================================================================
        //
        public override void SetProperty(string PropertyName, string Value, int TargetMemberId = 0) {
            if (TargetMemberId == 0) {
                CP.core.userProperty.setProperty(PropertyName, Value);
            } else {
                CP.core.userProperty.setProperty(PropertyName, Value, TargetMemberId);
            }
        }
        //
        //=======================================================================================================
        // todo REFACTOR -- obsolete this and setup the defaultValue type correctly, and add the targetUserId
        //=======================================================================================================
        //
        public override bool GetBoolean(string PropertyName, string DefaultValue = "") {
            return CP.core.userProperty.getBoolean(PropertyName, genericController.encodeBoolean(DefaultValue));
        }
        //
        //=======================================================================================================
        //  todo REFACTOR -- obsolete this and setup the defaultValue type correctly, and add the targetUserId
        //=======================================================================================================
        //
        public override DateTime GetDate(string PropertyName, string DefaultValue = "") {
            return CP.core.userProperty.getDate(PropertyName, genericController.encodeDate(DefaultValue));
        }
        //
        //=======================================================================================================
        // todo  REFACTOR -- obsolete this and setup the defaultValue type correctly, and add the targetUserId
        //=======================================================================================================
        //
        public override int GetInteger(string PropertyName, string DefaultValue = "") {
            return CP.core.userProperty.getInteger(PropertyName, genericController.encodeInteger(DefaultValue));
        }
        //
        //=======================================================================================================
        // todo  REFACTOR -- obsolete this and setup the defaultValue type correctly, and add the targetUserId
        //=======================================================================================================
        //
        public override double GetNumber(string PropertyName, string DefaultValue = "") {
            return CP.core.userProperty.getNumber(PropertyName, encodeNumber(DefaultValue));
        }
        //
        //=======================================================================================================
        //
        public override string GetText(string PropertyName, string DefaultValue = "") {
            return CP.core.userProperty.getText(PropertyName, DefaultValue);
        }
        //
        //====================================================================================================
        // todo  obsolete
        //
        public override void Track() {
            int localId = Id;
        }
        //
        //====================================================================================================
        //
        private void appendDebugLog(string copy) {
            //
        }
        //
        //====================================================================================================
        //
        private void tp(string msg) {}
        //
        //====================================================================================================
        //
        #region  IDisposable Support 
        // Do not change or add Overridable to these methods.
        // Put cleanup code in Dispose(ByVal disposing As Boolean).
        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~CPUserClass() {
            Dispose(false);
            //todo  NOTE: The base class Finalize method is automatically called from the destructor:
            //base.Finalize();
        }
        #endregion
    }

}